### DISCUSSION TITLE
How to  estimate depletion voltage for double-sided  orthogonal  strip detector using SSD

### DISCUSSION THREAD

**Discussion by zql2021** [JuliaPhysics/SolidStateDetectors.jl#386] [Q&A] [answered]
Dear Dr. Hagemann
I hope I’m not interrupting at an inconvenient time. I am currently encountering some issues with estimating the depletion voltage for double-sided orthogonal strip detectors using SSD.
1. I found that when  bias voltage is applied to multiple electrodes, `estimate_depletion_voltage()` couldn't work. I set **all the top electrodes** bias voltage to 1200V in detconfig.yaml, resulting in the error: `'ERROR: AssertionError: Could not determine contact at which the bias voltage is applied as multiple contacts have non-zero contact potentials.'"`
2. If I set the voltage to 1200V for only one strip electrode, and I got:
```
[ Info: Looking for the depletion voltage applied to contact 6 in the range (0.0f0 V, 1200.0f0 V).
[ Info: The depletion voltage is around 1199.9 ± 0.1 V applied to contact 6.
┌ Warning: The depletion voltage is probably not in the specified range (0.0f0 V, 1200.0f0 V).
└ @ SolidStateDetectors ~/.julia/packages/SolidStateDetectors/rx7KB/src/Simulation/DepletionVoltage.jl:111
1199.9268f0 V
```
I am estimating the depletion voltage to verify if there are any errors in detconfig.yaml. Additionally, I performed some tests: executing `is_depleted(sim_points_types) `returned true, and `get_active_volume(sim.point_type) `returned 11.88 cm³, both of them seem ok.

Could you provide some suggestions or enhancements on estimating the depletion voltage for a double-sided orthogonal strip detector using SSD? Setting the potential of all top electrodes to 1200V in **detconfig.yaml** may be too high. I would like to **determine the potential setting** based on the estimated depletion voltage.

Below is my detconfig.yaml, hoping it will help you better understand my issue.
[detconfig.zip](https://github.com/user-attachments/files/16039012/detconfig.zip)
Looking forward to your reply, Thanks a lot.

**Comment by fhagemann** [comment, marked as answer]
Hi,
as you pointed out, the current implementation of estimating the depletion voltage assumes that the non-zero potential is only applied to one of the contacts.
There are some tweaks you could use to modify the detector such that you can use the SSD function:

1. Combine all contacts with potential 1200V into ONE contact, and all contacts with potential 0V into ANOTHER contact, so you end up with a detector that "consists" of only two contacts. Note that the following code is hard-coded and only works if the biased contacts are at potential 1200V and if the first contact is a biased contact and the second one is not:
    ```julia
    using SolidStateDetectors
    T = Float64

    # read in the configuration file as dictionary
    config_dict = SolidStateDetectors.parse_config_file("detconfig.yaml")

    # determine the indices of all contacts that have a potential of 1200 (and the ones who do not)
    potentials = getindex.(config_dict["detectors"][1]["contacts"], "potential")
    biased = findall(potentials .== 1200)
    notbiased = findall(potentials .!= 1200)

    # combine all geometries of biased contacts into one
    biasedgeometries = getindex.(config_dict["detectors"][1]["contacts"][biased], "geometry")
    config_dict["detectors"][1]["contacts"][1]["geometry"] = Dict("union" => biasedgeometries)

    # combine all geometries of unbiased contacts into one
    notbiasedgeometries = getindex.(config_dict["detectors"][1]["contacts"][notbiased], "geometry")
    config_dict["detectors"][1]["contacts"][2]["geometry"] = Dict("union" => notbiasedgeometries)

    # only keep the first two contacts and delete all other contacts
    config_dict["detectors"][1]["contacts"] = config_dict["detectors"][1]["contacts"][1:2]

    sim = Simulation{T}(config_dict)
    ```
    ![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/30291312/2b423c6a-f39a-4920-8f81-2fa4db45b257)

    Then, you could use `estimate_depletion_voltage` to get the potential applied to contact 1, needed to deplete the detector.
     ```julia
    calculate_electric_potential!(sim, refinement_limits = [0.2,0.1,0.05,0.03,0.02], depletion_handling = true)
    calculate_weighting_potential!(sim, 1, refinement_limits = [0.2,0.1,0.05,0.03,0.02])
    estimate_depletion_voltage(sim)
    ```
    ```
    [ Info: Looking for the depletion voltage applied to contact 1 in the range (0.0 V, 1200.0 V).
    [ Info: The depletion voltage is around 1047.9 ± 0.1 V applied to contact 1.
    1047.8759765625 V
    ```

2. Basically what you have is a cubic detector. You could simulate the setup faster by approximating the top contact to cover the whole top surface and the bottom contact to cover the whole bottom surface. In the configuration file, this would be:
    ```yaml
    contacts:
      - name: ring-top
        material: HPGe 
        potential: 1200
        id: 1
        geometry:
          box:
            widths: [26.0, 26.0, 0.01]
            origin: [0, 0, 15.005]
      - name: ring-bottom
        material: HPGe 
        potential: 0
        id: 2
        geometry:
          box:
            widths: [26.0, 26.0, 0.01]
            origin: [0, 0, -0.005]
   ```
    ![image](https://github.com/JuliaPhysics/SolidStateDetectors.jl/assets/30291312/02242777-af8e-4b12-962b-e13336ddbafe)
    In this case, you would get 
    ```julia
    calculate_electric_potential!(sim, refinement_limits = [0.2,0.1,0.05,0.03,0.02], depletion_handling = true)
    calculate_weighting_potential!(sim, 1, refinement_limits = [0.2,0.1,0.05,0.03,0.02])
    estimate_depletion_voltage(sim)
    ```
    ```
    [ Info: Looking for the depletion voltage applied to contact 1 in the range (0.0 V, 1200.0 V).
    [ Info: The depletion voltage is around 1043.0 ± 0.1 V applied to contact 1.
    1043.0419921875 V
    ```

Both methods return quite similar results, so you can assume that your detector is fully depleted already at 1050V.

**Comment by fhagemann** [comment]
Of course, both of these methods require to recalculate the electric potential and weighting potentials.
How important would it be for you to have a function that could calculate this without having to modify the detector object?
@zql2021 
We could have a look into using already calculated weighting potentials, but this will probably require some time to incorporate and thoroughly test, so I cannot provide anything to be used right now..

**Reply by zql2021** [reply]
Thank you very much for your patient reply. I just want to roughly estimate the depletion voltage of my detector so I can determine `potential` in detconfig.yaml and continue simulating waveforms. I believe the first method, "Combine all contacts with potential 1200V into ONE contact" would solve my problem, so there's no need for a specialized function to calculate this without modifying the detector object.
Thanks again for you help!
